default_gateway.go

```diff
@@ -0,0 +1,158 @@
+   endpoints have GW set by the driver. ICC is set to false for
+   the GW_bridge network.
+
+   If a driver can't provide external connectivity it can choose
+   to not set the GW IP for the endpoint.
+
+   endpoint on the GW_bridge network is managed dynamically by
+   libnetwork. ie:
+   - its created when an endpoint without GW joins the container
+   - its deleted when an endpoint with GW joins the container
+*/
+
+func (sb *sandbox) setupDefaultGW(srcEp *endpoint) error {
+	var createOptions []EndpointOption
+	c := srcEp.network.ctrlr
+
+	// check if the conitainer already has a GW endpoint
+	if ep := sb.getEndpointInGWNetwork(); ep != nil {
+		return nil
+	}
+
+	n, err := c.NetworkByName(libnGWNetwork)
+	if err != nil {
+		if _, ok := err.(ErrNoSuchNetwork); !ok {
+			return err
+		}
+		n, err = createGWNetwork(c)
+		if err != nil {
+			return err
+		}
+	}
+
+	if opt, ok := srcEp.generic[netlabel.PortMap]; ok {
+		if pb, ok := opt.([]types.PortBinding); ok {
+			createOptions = append(createOptions, CreateOptionPortMapping(pb))
+		}
+	}
+
+	if opt, ok := srcEp.generic[netlabel.ExposedPorts]; ok {
+		if exp, ok := opt.([]types.TransportPort); ok {
+			createOptions = append(createOptions, CreateOptionExposedPorts(exp))
+		}
+	}
+
+	newEp, err := n.CreateEndpoint(srcEp.name, createOptions...)
+	if err != nil {
+		return fmt.Errorf("Endpoint create on GW Network failed: %v", err)
+	}
+
+	if err := newEp.Join(sb); err != nil {
+		return fmt.Errorf("Endpoint join on GW Network failed: %v", err)
+	}
+	return nil
+}
+
+func (sb *sandbox) clearDefaultGW() error {
+	var ep *endpoint
+
+	if ep = sb.getEndpointInGWNetwork(); ep == nil {
+		return nil
+	}
+
+	if err := ep.Leave(sb); err != nil {
+		return fmt.Errorf("Endpoint leaving GW Network failed: %v", err)
+	}
+	if err := ep.Delete(); err != nil {
+		return fmt.Errorf("Deleting endpoint on GW Network failed: %v", err)
+	}
+	return nil
+}
+
+func createGWNetwork(c NetworkController) (Network, error) {
+	netOption := options.Generic{
+		"BridgeName":            libnGWNetwork,
+		"EnableICC":             false,
+		"AllowNonDefaultBridge": true,
+		"EnableIPMasquerade":    true,
+	}
+
+	n, err := c.NewNetwork("bridge", libnGWNetwork,
+		NetworkOptionGeneric(options.Generic{
+			netlabel.GenericData: netOption,
+			netlabel.EnableIPv6:  false,
+		}))
+
+	if err != nil {
+		return nil, fmt.Errorf("Error creating external connectivity network: %v", err)
+	}
+	return n, err
+}
+
+func (sb *sandbox) needDefaultGW() bool {
+	var validep bool
+	eps := sb.getConnectedEndpoints()
+
+	for _, ep := range eps {
+		if ep.endpointInGWNetwork() {
+			continue
+		}
+		if ep.network.networkType == "null" || ep.network.networkType == "host" {
+			continue
+		}
+		// TODO v6 needs to be handled.
+		if ep.joinInfo.gw != nil {
+			return false
+		}
+		validep = true
+	}
+	if validep {
+		return true
+	}
+	return false
+}
+
+func (sb *sandbox) getEndpointInGWNetwork() *endpoint {
```

Comment:
Where this function is being used, it is being used only as a true/false switch semantic.
I think we should change this function to just return a bool and change the name accordingly.

Probably you needed for a previously, but with current algorithm it is no longer needed.